# Groupe 5

## Text Mining

### Franck SORBY, Lorenzo AGUILERA, Johan BOUSQUIE

# Etape 1 :
## Import des librairies

In [94]:
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import CountVectorizer
import pandas as pd
import re
import string
import nltk

stopwords = nltk.corpus.stopwords.words('english')
ps = nltk.PorterStemmer()

# Ouverture document

In [95]:
data = pd.read_csv("https://gist.githubusercontent.com/aguileralorenzo/a455ed7b55c3393aaf3867758bbb5904/raw/1af9811cff325deb604e5a81c32be1dc02101cb0/spam_collection.txt", sep='\t', header=None)
data.columns = ['label','msg']

# Split en train et test

In [96]:
X_train, X_test, Y_train, Y_test = train_test_split(data['msg'],data['label'],shuffle=False)

vectorizer = CountVectorizer()
X = vectorizer.fit_transform(X_train)

clf = LogisticRegression()
clf.fit(X,Y_train)

clf.score(X,Y_train)

/Users/joelebonobo/opt/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
/Users/joelebonobo/opt/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:469: FutureWarning: Default multi_class will be changed to 'auto' in 0.22. Specify the multi_class option to silence this warning.
  "this warning.", FutureWarning)


0.9976070830342187

# Premier score 
Nous avons un score de 0.9976070830342187 qui risque d'être difficile à améliorer cepandant nous allons quand
meme utiliser le grid search pour essayer d'améliorer le score.

In [48]:
params = {'C': [0.01,0.1,1,5,10,15,20], 'fit_intercept':[True,False]}

Ici nous faisons un premier GS pour se rapprocher des meilleurs parametres
puis nous explorerons plus en profondeur

In [49]:
from sklearn.model_selection import GridSearchCV
gs=GridSearchCV(clf,params, refit=True).fit(X,Y_train)

/Users/joelebonobo/opt/anaconda3/lib/python3.7/site-packages/sklearn/model_selection/_split.py:1978: FutureWarning: The default value of cv will change from 3 to 5 in version 0.22. Specify it explicitly to silence this warning.
  warnings.warn(CV_WARNING, FutureWarning)
/Users/joelebonobo/opt/anaconda3/lib/python3.7/site-packages/sklearn/model_selection/_split.py:657: Warning: The least populated class in y has only 1 members, which is too few. The minimum number of members in any class cannot be less than n_splits=3.
  % (min_groups, self.n_splits)), Warning)
/Users/joelebonobo/opt/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
/Users/joelebonobo/opt/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:469: FutureWarning: Default multi_class will be changed to 'auto' in 0.22. Specify the multi_class option to silenc

In [50]:
gs.best_params_

{'C': 15, 'fit_intercept': True}

# Approfondissement
Nous allons donc chercher autour de 15

In [51]:
params = {'C': [10,11,12,13,14,15,16,17,18,19,20], 'fit_intercept':[True,False]}
from sklearn.model_selection import GridSearchCV
gs=GridSearchCV(clf,params, refit=True).fit(X,Y_train)
gs.best_params_

/Users/joelebonobo/opt/anaconda3/lib/python3.7/site-packages/sklearn/model_selection/_split.py:1978: FutureWarning: The default value of cv will change from 3 to 5 in version 0.22. Specify it explicitly to silence this warning.
  warnings.warn(CV_WARNING, FutureWarning)
/Users/joelebonobo/opt/anaconda3/lib/python3.7/site-packages/sklearn/model_selection/_split.py:657: Warning: The least populated class in y has only 1 members, which is too few. The minimum number of members in any class cannot be less than n_splits=3.
  % (min_groups, self.n_splits)), Warning)
/Users/joelebonobo/opt/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
/Users/joelebonobo/opt/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:469: FutureWarning: Default multi_class will be changed to 'auto' in 0.22. Specify the multi_class option to silenc

{'C': 12, 'fit_intercept': True}

In [52]:
params = {'C': [11,11.1,11.2,11.3,11.4,11.5,11.6,11.7,11.8,11.9,12,12.1,12.2,12.3,12.4,12.5,12.6,12.7,12.8,12.9,13], 'fit_intercept':[True,False]}
from sklearn.model_selection import GridSearchCV
gs=GridSearchCV(clf,params, refit=True).fit(X,Y_train)
gs.best_params_

/Users/joelebonobo/opt/anaconda3/lib/python3.7/site-packages/sklearn/model_selection/_split.py:1978: FutureWarning: The default value of cv will change from 3 to 5 in version 0.22. Specify it explicitly to silence this warning.
  warnings.warn(CV_WARNING, FutureWarning)
/Users/joelebonobo/opt/anaconda3/lib/python3.7/site-packages/sklearn/model_selection/_split.py:657: Warning: The least populated class in y has only 1 members, which is too few. The minimum number of members in any class cannot be less than n_splits=3.
  % (min_groups, self.n_splits)), Warning)
/Users/joelebonobo/opt/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
/Users/joelebonobo/opt/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:469: FutureWarning: Default multi_class will be changed to 'auto' in 0.22. Specify the multi_class option to silenc

{'C': 11.7, 'fit_intercept': True}

In [53]:
params = {'C': [11.65,11.68,11.7,11.72,11.74,11.76], 'fit_intercept':[True,False]}
from sklearn.model_selection import GridSearchCV
gs=GridSearchCV(clf,params, refit=True).fit(X,Y_train)
gs.best_params_

/Users/joelebonobo/opt/anaconda3/lib/python3.7/site-packages/sklearn/model_selection/_split.py:1978: FutureWarning: The default value of cv will change from 3 to 5 in version 0.22. Specify it explicitly to silence this warning.
  warnings.warn(CV_WARNING, FutureWarning)
/Users/joelebonobo/opt/anaconda3/lib/python3.7/site-packages/sklearn/model_selection/_split.py:657: Warning: The least populated class in y has only 1 members, which is too few. The minimum number of members in any class cannot be less than n_splits=3.
  % (min_groups, self.n_splits)), Warning)
/Users/joelebonobo/opt/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
/Users/joelebonobo/opt/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:469: FutureWarning: Default multi_class will be changed to 'auto' in 0.22. Specify the multi_class option to silenc

{'C': 11.65, 'fit_intercept': True}

In [54]:
params = {'C': [11.61,11.62,11.63,11.64,11.65,11.66,11.67], 'fit_intercept':[True,False]}
from sklearn.model_selection import GridSearchCV
gs=GridSearchCV(clf,params, refit=True).fit(X,Y_train)
gs.best_params_

/Users/joelebonobo/opt/anaconda3/lib/python3.7/site-packages/sklearn/model_selection/_split.py:1978: FutureWarning: The default value of cv will change from 3 to 5 in version 0.22. Specify it explicitly to silence this warning.
  warnings.warn(CV_WARNING, FutureWarning)
/Users/joelebonobo/opt/anaconda3/lib/python3.7/site-packages/sklearn/model_selection/_split.py:657: Warning: The least populated class in y has only 1 members, which is too few. The minimum number of members in any class cannot be less than n_splits=3.
  % (min_groups, self.n_splits)), Warning)
/Users/joelebonobo/opt/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
/Users/joelebonobo/opt/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:469: FutureWarning: Default multi_class will be changed to 'auto' in 0.22. Specify the multi_class option to silenc

{'C': 11.61, 'fit_intercept': True}

In [55]:
params = {'C': [11.6,11.61,11.69,11.7], 'fit_intercept':[True,False]}
from sklearn.model_selection import GridSearchCV
gs=GridSearchCV(clf,params, refit=True).fit(X,Y_train)
gs.best_params_

/Users/joelebonobo/opt/anaconda3/lib/python3.7/site-packages/sklearn/model_selection/_split.py:1978: FutureWarning: The default value of cv will change from 3 to 5 in version 0.22. Specify it explicitly to silence this warning.
  warnings.warn(CV_WARNING, FutureWarning)
/Users/joelebonobo/opt/anaconda3/lib/python3.7/site-packages/sklearn/model_selection/_split.py:657: Warning: The least populated class in y has only 1 members, which is too few. The minimum number of members in any class cannot be less than n_splits=3.
  % (min_groups, self.n_splits)), Warning)
/Users/joelebonobo/opt/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
/Users/joelebonobo/opt/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:469: FutureWarning: Default multi_class will be changed to 'auto' in 0.22. Specify the multi_class option to silenc

{'C': 11.61, 'fit_intercept': True}

In [62]:
params = {'C': [11.609,11.6081,11.6082,11.6083,11.6084,11.6085,11.6086,11.6087,11.6088,11.6089,11.61], 'fit_intercept':[True,False]}
from sklearn.model_selection import GridSearchCV
gs=GridSearchCV(clf,params, refit=True).fit(X,Y_train)
gs.best_params_

/Users/joelebonobo/opt/anaconda3/lib/python3.7/site-packages/sklearn/model_selection/_split.py:1978: FutureWarning: The default value of cv will change from 3 to 5 in version 0.22. Specify it explicitly to silence this warning.
  warnings.warn(CV_WARNING, FutureWarning)
/Users/joelebonobo/opt/anaconda3/lib/python3.7/site-packages/sklearn/model_selection/_split.py:657: Warning: The least populated class in y has only 1 members, which is too few. The minimum number of members in any class cannot be less than n_splits=3.
  % (min_groups, self.n_splits)), Warning)
/Users/joelebonobo/opt/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
/Users/joelebonobo/opt/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:469: FutureWarning: Default multi_class will be changed to 'auto' in 0.22. Specify the multi_class option to silenc

{'C': 11.609, 'fit_intercept': True}

# Best Param
Ici nous pouvons voir que le meilleur paramètre est 11,609

In [90]:
clf = LogisticRegression(C=11.609)
clf.fit(X,Y_train)

clf.score(X,Y_train)

0.9995214166068438

# Best Parameters
Le meilleur parametre pour C pour la LogisticRegression semble être 11.609
après de nombreux essais nous avons pu augmenter notre score 
 bien qu'il était déja élevé.
  0.9995214166068438 > 0.9976070830342187

# Deuxieme document
## 20 news groups

Import de nos documents splités, train and test
Ici nous prenons que 4 catécogries


In [98]:
from sklearn.datasets import fetch_20newsgroups

from sklearn.feature_extraction.text import TfidfVectorizer
categories = ['alt.atheism', 'talk.religion.misc',
              'comp.graphics', 'sci.space']
newsgroups_train = fetch_20newsgroups(subset='train',
                                      categories=categories)


# Premier score

In [100]:
vectorizer = TfidfVectorizer()
vectors = vectorizer.fit_transform(newsgroups_train.data)
vectors.shape

vectors.nnz / float(vectors.shape[0])

from sklearn.naive_bayes import MultinomialNB
from sklearn import metrics
newsgroups_test = fetch_20newsgroups(subset='test',
                                     categories=categories)
vectors_test = vectorizer.transform(newsgroups_test.data)
clf = LogisticRegression()
clf.fit(vectors, newsgroups_train.target)
pred = clf.predict(vectors_test)
metrics.f1_score(newsgroups_test.target, pred, average='macro')


/Users/joelebonobo/opt/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
/Users/joelebonobo/opt/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:469: FutureWarning: Default multi_class will be changed to 'auto' in 0.22. Specify the multi_class option to silence this warning.
  "this warning.", FutureWarning)


0.8511363057725347

### Le premier score est de 0.851

Nous allons essayer de l'améliorer à l'aide du GS

## Recherche du best Param avec GridSearchCV

In [105]:
params_20news = {'C': [0.01,0.1,1,10,15,20], 'fit_intercept':[True,False]}
from sklearn.model_selection import GridSearchCV
gs2=GridSearchCV(clf,params_20news, refit=True).fit(X,Y_train)
gs2.best_params_

/Users/joelebonobo/opt/anaconda3/lib/python3.7/site-packages/sklearn/model_selection/_split.py:1978: FutureWarning: The default value of cv will change from 3 to 5 in version 0.22. Specify it explicitly to silence this warning.
  warnings.warn(CV_WARNING, FutureWarning)
/Users/joelebonobo/opt/anaconda3/lib/python3.7/site-packages/sklearn/model_selection/_split.py:657: Warning: The least populated class in y has only 1 members, which is too few. The minimum number of members in any class cannot be less than n_splits=3.
  % (min_groups, self.n_splits)), Warning)
/Users/joelebonobo/opt/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
/Users/joelebonobo/opt/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:469: FutureWarning: Default multi_class will be changed to 'auto' in 0.22. Specify the multi_class option to silenc

{'C': 15, 'fit_intercept': True}

In [106]:
params_20news = {'C': [10,12,14], 'fit_intercept':[True,False]}
from sklearn.model_selection import GridSearchCV
gs2=GridSearchCV(clf,params_20news, refit=True).fit(X,Y_train)
gs2.best_params_

/Users/joelebonobo/opt/anaconda3/lib/python3.7/site-packages/sklearn/model_selection/_split.py:1978: FutureWarning: The default value of cv will change from 3 to 5 in version 0.22. Specify it explicitly to silence this warning.
  warnings.warn(CV_WARNING, FutureWarning)
/Users/joelebonobo/opt/anaconda3/lib/python3.7/site-packages/sklearn/model_selection/_split.py:657: Warning: The least populated class in y has only 1 members, which is too few. The minimum number of members in any class cannot be less than n_splits=3.
  % (min_groups, self.n_splits)), Warning)
/Users/joelebonobo/opt/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
/Users/joelebonobo/opt/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:469: FutureWarning: Default multi_class will be changed to 'auto' in 0.22. Specify the multi_class option to silenc

{'C': 12, 'fit_intercept': True}

# Approfondissement

In [107]:
params_20news = {'C': [11.5,12,12.5], 'fit_intercept':[True,False]}
from sklearn.model_selection import GridSearchCV
gs2=GridSearchCV(clf,params_20news, refit=True).fit(X,Y_train)
gs2.best_params_

/Users/joelebonobo/opt/anaconda3/lib/python3.7/site-packages/sklearn/model_selection/_split.py:1978: FutureWarning: The default value of cv will change from 3 to 5 in version 0.22. Specify it explicitly to silence this warning.
  warnings.warn(CV_WARNING, FutureWarning)
/Users/joelebonobo/opt/anaconda3/lib/python3.7/site-packages/sklearn/model_selection/_split.py:657: Warning: The least populated class in y has only 1 members, which is too few. The minimum number of members in any class cannot be less than n_splits=3.
  % (min_groups, self.n_splits)), Warning)
/Users/joelebonobo/opt/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
/Users/joelebonobo/opt/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:469: FutureWarning: Default multi_class will be changed to 'auto' in 0.22. Specify the multi_class option to silenc

{'C': 12, 'fit_intercept': True}

Nous allons garder 12 comme paramètre pour C dans la fonction Logistic Regression

# Test 

In [108]:
clf = LogisticRegression(C=12)
clf.fit(vectors, newsgroups_train.target)
pred = clf.predict(vectors_test)
metrics.f1_score(newsgroups_test.target, pred, average='macro')

0.873591439082844

0.873591439082844 > 0.8511363057725347
Nous avons donc réussi à améliorer le score
